pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/MethodNamingConventionsRule.java

```diff
@@ -4,63 +4,119 @@
+    }
+
 
-        if (Character.isUpperCase(methodName.charAt(0))) {
-            addViolationWithMessage(data, node, "Method names should not start with capital letters");
+    private boolean isJunit3Test(ASTMethodDeclaration node) {
+        if (!node.getMethodName().startsWith("test")) {
+            return false;
         }
-        if (methodName.indexOf('_') >= 0) {
-            addViolationWithMessage(data, node, "Method names should not contain underscores");
+
+        // Considers anonymous classes, TODO with #905 this will be easier
+        Node parent = node.getFirstParentOfAnyType(ASTEnumConstant.class, ASTAllocationExpression.class, ASTAnyTypeDeclaration.class);
+
+        if (!(parent instanceof ASTClassOrInterfaceDeclaration) || ((ASTClassOrInterfaceDeclaration) parent).isInterface()) {
+            return false;
         }
-        return data;
+
+        ASTClassOrInterfaceType superClass = ((ASTClassOrInterfaceDeclaration) parent).getSuperClassTypeNode();
+
+        return superClass != null && TypeHelper.isA(superClass, "junit.framework.TestCase");
     }
 
-    private boolean isOverriddenMethod(ASTMethodDeclarator node) {
-        ASTClassOrInterfaceBodyDeclaration declaration = node
-                .getFirstParentOfType(ASTClassOrInterfaceBodyDeclaration.class);
-        List<ASTMarkerAnnotation> annotations = declaration.findDescendantsOfType(ASTMarkerAnnotation.class);
-        for (ASTMarkerAnnotation ann : annotations) {
-            ASTName name = ann.getFirstChildOfType(ASTName.class);
-            if (name != null && name.hasImageEqualTo("Override")) {
-                return true;
+
+    @Override
+    public Object visit(ASTMethodDeclaration node, Object data) {
+
+        if (node.isAnnotationPresent("java.lang.Override")) {
+            return super.visit(node, data);
+        }
+
+        if (node.isNative()) {
+            if (getProperty(CHECK_NATIVE_METHODS_DESCRIPTOR)) {
+                checkMatches(node, NATIVE_REGEX, data);
+            } else {
+                return super.visit(node, data);
             }
+        } else if (node.isStatic()) {
+            checkMatches(node, STATIC_REGEX, data);
+        } else if (isJunit4Test(node)) {
+            checkMatches(node, JUNIT4_REGEX, data);
+        } else if (isJunit3Test(node)) {
+            checkMatches(node, JUNIT3_REGEX, data);
+        } else {
+            checkMatches(node, INSTANCE_REGEX, data);
         }
-        return false;
+
+        return super.visit(node, data);
+    }
+
+
+    private static RegexPBuilder defaultProp(String displayName) {
+        String propName = StringUtil.toCamelCase(displayName, true) + "Pattern";
+        DESCRIPTOR_TO_DISPLAY_NAME.put(propName, displayName);
+
+        return RegexProperty.named(propName)
+                            .desc("Regex which applies to " + displayName.trim() + " method names")
+                            .defaultValue("[a-z][a-zA-Z0-9]+");
```

Comment:
Maybe we should use "Predefined Character classes" for this default pattern:

`(?U)\p{JavaIdentifierStart}\p{JavaIdentifierPart}+`

That would not start with a lower-case letter. So, maybe this one:

`(?U)[\p{Alpha}&&[\p{Lower}]]\p{JavaIdentifierPart}+`

Note: I didn't try these out... Just reading the documentation https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html

Hm... well, maybe that's too much. I think, I usually don't use other characters than `[a-z][a-zA-Z0-9]+`....

Reply:
I didn't really think about unicode support, but they might be in use somewhere. On the one hand, restricting the alphabet to english characters could be seen as gatekeeping by the people that want additional characters. On the other hand, most people code in english, and it makes sense to disallow non-english characters (and `_$`) to keep identifiers clean... I think I'd rather keep it simple and trust the people that use additionnal characters to define themselves which characters they want to use and which ones they want to disallow.  

I tried out `\p{JavaIdentifierPart}` in jshell and it's not recognised. Are you sure that's a thing?

Reply:
Yes, it is configurable by the end users, so we keep it. It's a sensible default value.

According to the doc, `\p{JavaIdentifierPart}` should exist (it is actually mapped to Character.isJavaIdentifierPart(char)). It might need to be enabled with `(?U)`... no - now I understand the documentation, after reading the source. The correct usage is `\p{javaJavaIdentifierPart}` (lowercase java first to point to the Character methods, then JavaIdentifierPart - which is the method name).